include/ignition/math/detail/Ellipsoid.hh

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef IGNITION_MATH_DETAIL_CAPSULE_HH_
+#define IGNITION_MATH_DETAIL_CAPSULE_HH_
+
+#include <limits>
+#include <optional>
+#include <ignition/math/Helpers.hh>
+#include <ignition/math/Inertial.hh>
+
+namespace ignition
+{
+namespace math
+{
+
+//////////////////////////////////////////////////
+template<typename T>
+Ellipsoid<T>::Ellipsoid(const Vector3<T> &_radii) : radii(_radii) {}
+
+//////////////////////////////////////////////////
+template<typename T>
+Ellipsoid<T>::Ellipsoid(const Vector3<T> &_radii, const Material &_mat)
+: radii(_radii), material(_mat) {}
+
+//////////////////////////////////////////////////
+template<typename T>
+Vector3<T> Ellipsoid<T>::Radii() const
+{
+  return this->radii;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+void Ellipsoid<T>::SetRadii(const Vector3<T> &_radii)
+{
+  this->radii = _radii;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+const Material &Ellipsoid<T>::Mat() const
+{
+  return this->material;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+void Ellipsoid<T>::SetMat(const Material &_mat)
+{
+  this->material = _mat;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+bool Ellipsoid<T>::operator==(const Ellipsoid &_ellipsoid) const
+{
+  return this->radii == _ellipsoid.radii &&
+    this->material == _ellipsoid.material;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+std::optional< MassMatrix3<T> > Ellipsoid<T>::MassMatrix() const
+{
+  if (this->radii.X() < 0 || this->radii.Y() < 0 || this->radii.Z() < 0)
```

Comment:
I think we want to use `<=` for these comparisons, since the inertial properties are invalid if any radii are `0`

Reply:
Completely reasonable and fixed, but then that begs another question. If 0 and negative values are invalid, should they be checked during construction/`SetRadii()`? And an exception thrown that case?

Reply:
> If 0 and negative values are invalid, should they be checked during construction/`SetRadii()`? And an exception thrown in that case?

Currently the constructors and Get/Set methods act like plain old data structures, and there's only validity checking in the methods dealing with inertia computations. I'm inclined to keep it that way for `ign-math6` so that the shape classes have the same exception behavior. I think we could consider that for `ignition-math7`, but it would be good to consider the impact it would have on downstream code in libsdformat, ign-gazebo, etc.